File: lib/emails.js
import { consoleProgress } from "./progress.js";

const EMAIL_RE = /[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}/g;

const PATHS = [
  "",
  "/contato",
  "/contatos",
  "/contact",
  "/contact-us",
  "/sobre",
  "/about",
  "/fale-conosco",
  "/equipe",
  "/quem-somos",
];

const PLAUSIBLE_TLDS = new Set([
  "com",
  "br",
  "net",
  "org",
  "io",
  "co",
  "dev",
  "app",
  "tech",
  "ag",
  "agency",
  "studio",
  "design",
  "digital",
  "me",
  "pt",
  "eu",
  "us",
  "info",
  "biz",
  "tv",
  "cc",
  "xyz",
  "ai",
  "gg",
  "page",
  "site",
]);

const BLOCKLIST = [
  "sentry.io",
  "sentry-next.wixpress.com",
  "wixpress.com",
  "example.com",
  "domain.com",
  "email.com",
  "seudominio.com",
  "yourdomain.com",
  "test.com",
  "mysite.com",
  "website.com",
];

const FETCH_TIMEOUT_MS = 8000;

async function fetchText(url) {
  const ctrl = new AbortController();
  const t = setTimeout(() => ctrl.abort(), FETCH_TIMEOUT_MS);
  try {
    const res = await fetch(url, {
      signal: ctrl.signal,
      redirect: "follow",
      headers: {
        "User-Agent":
          "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36",
        "Accept-Language": "pt-BR,pt;q=0.9,en;q=0.5",
      },
    });
    if (!res.ok) return "";
    const ct = res.headers.get("content-type") || "";
    if (!ct.includes("text") && !ct.includes("html")) return "";
    return await res.text();
  } catch {
    return "";
  } finally {
    clearTimeout(t);
  }
}

function preprocess(html) {
  let s = html
    .replace(/&#64;/gi, "@")
    .replace(/&#0?46;/gi, ".")
    .replace(/&commat;/gi, "@")
    .replace(/&period;/gi, ".")
    .replace(/&amp;/gi, "&")
    .replace(/&nbsp;/gi, " ")
    .replace(/[​-‍﻿­]/g, "")
    .replace(/<wbr\s*\/?>/gi, "")
    .replace(/<!--.*?-->/gs, "");

  s = s
    .replace(/\s*\[\s*at\s*\]\s*/gi, "@")
    .replace(/\s*\(\s*at\s*\)\s*/gi, "@")
    .replace(/\s+at\s+(?=[a-z0-9-]+\s*(?:\[|\()\s*dot)/gi, "@")
    .replace(/\s*\(arroba\)\s*/gi, "@")
    .replace(/\s+arroba\s+/gi, "@")
    .replace(/\s*\[\s*dot\s*\]\s*/gi, ".")
    .replace(/\s*\(\s*dot\s*\)\s*/gi, ".")
    .replace(/\s*\(ponto\)\s*/gi, ".")
    .replace(/\s+ponto\s+/gi, ".");

  return s;
}

function tldOf(email) {
  const parts = email.split("@")[1]?.split(".") || [];
  return parts[parts.length - 1]?.toLowerCase() || "";
}

function rootDomain(host) {
  return host.replace(/^www\./, "").toLowerCase();
}

function isValidEmail(email, siteHost) {
  if (BLOCKLIST.some((b) => email.includes(b))) return false;
  if (/\.(png|jpe?g|gif|svg|webp|woff2?|ttf|ico|css|js)(\?|$)/.test(email))
    return false;
  if (/^[a-f0-9]{16,}@/i.test(email)) return false;
  if (email.length > 80) return false;

  const tld = tldOf(email);
  if (PLAUSIBLE_TLDS.has(tld)) return true;

  if (siteHost) {
    const siteTld = rootDomain(siteHost).split(".").pop();
    if (tld === siteTld) return true;
  }
  return false;
}

function extract(html, siteHost) {
  const processed = preprocess(html);
  const mailtos = new Set();
  const plain = new Set();

  for (const m of processed.matchAll(/mailto:([^"'?\s>&]+)/gi))
    mailtos.add(m[1]);
  for (const m of processed.matchAll(EMAIL_RE)) plain.add(m[0]);

  const clean = (e) => e.toLowerCase().replace(/[.,;:]+$/, "");
  const mailtoList = [...mailtos]
    .map(clean)
    .filter((e) => isValidEmail(e, siteHost));
  const plainList = [...plain]
    .map(clean)
    .filter((e) => isValidEmail(e, siteHost));

  const root = siteHost ? rootDomain(siteHost) : "";
  const sameDomain = (e) =>
    root && (e.endsWith("@" + root) || e.endsWith("." + root));

  return (
    mailtoList.find(sameDomain) ||
    plainList.find(sameDomain) ||
    mailtoList[0] ||
    plainList[0] ||
    ""
  );
}

async function findEmail(website) {
  if (!website) return "";
  let base;
  try {
    base = new URL(website);
  } catch {
    return "";
  }

  for (const path of PATHS) {
    const url = `${base.protocol}//${base.hostname}${path}`;
    const html = await fetchText(url);
    if (!html) continue;
    const email = extract(html, base.hostname);
    if (email) return email;
  }
  return "";
}

export async function enrichEmails(leads, onProgress = consoleProgress) {
  const out = [];
  for (let i = 0; i < leads.length; i++) {
    const l = leads[i];
    const email = await findEmail(l.website);
    onProgress({
      type: "item",
      index: i + 1,
      total: leads.length,
      name: l.name,
      status: email || "(sem email)",
    });
    out.push({ ...l, email });
  }
  return out;
}

File: python/emails.py
"""Extração de emails dos sites — equivalente a lib/emails.js."""

import re
from urllib.parse import urlparse

import httpx

from progress import console_progress

EMAIL_RE = re.compile(r"[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}")

PATHS = [
    "",
    "/contato",
    "/contatos",
    "/contact",
    "/contact-us",
    "/sobre",
    "/about",
    "/fale-conosco",
    "/equipe",
    "/quem-somos",
]

PLAUSIBLE_TLDS = {
    "com", "br", "net", "org", "io", "co", "dev", "app", "tech", "ag",
    "agency", "studio", "design", "digital", "me", "pt", "eu", "us",
    "info", "biz", "tv", "cc", "xyz", "ai", "gg", "page", "site",
}

BLOCKLIST = [
    "sentry.io",
    "sentry-next.wixpress.com",
    "wixpress.com",
    "example.com",
    "domain.com",
    "email.com",
    "seudominio.com",
    "yourdomain.com",
    "test.com",
    "mysite.com",
    "website.com",
]

FETCH_TIMEOUT_S = 8.0

_HEADERS = {
    "User-Agent": (
        "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 "
        "(KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36"
    ),
    "Accept-Language": "pt-BR,pt;q=0.9,en;q=0.5",
}


def fetch_text(url):
    try:
        with httpx.Client(
            timeout=FETCH_TIMEOUT_S, follow_redirects=True, headers=_HEADERS
        ) as client:
            res = client.get(url)
        if not res.is_success:
            return ""
        ct = res.headers.get("content-type", "")
        if "text" not in ct and "html" not in ct:
            return ""
        return res.text
    except Exception:
        return ""


def preprocess(html):
    s = html
    s = re.sub(r"&#64;", "@", s, flags=re.IGNORECASE)
    s = re.sub(r"&#0?46;", ".", s, flags=re.IGNORECASE)
    s = re.sub(r"&commat;", "@", s, flags=re.IGNORECASE)
    s = re.sub(r"&period;", ".", s, flags=re.IGNORECASE)
    s = re.sub(r"&amp;", "&", s, flags=re.IGNORECASE)
    s = re.sub(r"&nbsp;", " ", s, flags=re.IGNORECASE)
    s = re.sub(r"[​-‍﻿­]", "", s)
    s = re.sub(r"<wbr\s*/?>", "", s, flags=re.IGNORECASE)
    s = re.sub(r"<!--.*?-->", "", s, flags=re.DOTALL)

    s = re.sub(r"\s*\[\s*at\s*\]\s*", "@", s, flags=re.IGNORECASE)
    s = re.sub(r"\s*\(\s*at\s*\)\s*", "@", s, flags=re.IGNORECASE)
    s = re.sub(r"\s+at\s+(?=[a-z0-9-]+\s*(?:\[|\()\s*dot)", "@", s, flags=re.IGNORECASE)
    s = re.sub(r"\s*\(arroba\)\s*", "@", s, flags=re.IGNORECASE)
    s = re.sub(r"\s+arroba\s+", "@", s, flags=re.IGNORECASE)
    s = re.sub(r"\s*\[\s*dot\s*\]\s*", ".", s, flags=re.IGNORECASE)
    s = re.sub(r"\s*\(\s*dot\s*\)\s*", ".", s, flags=re.IGNORECASE)
    s = re.sub(r"\s*\(ponto\)\s*", ".", s, flags=re.IGNORECASE)
    s = re.sub(r"\s+ponto\s+", ".", s, flags=re.IGNORECASE)
    return s


def tld_of(email):
    parts = email.split("@")[1].split(".") if "@" in email else []
    return parts[-1].lower() if parts else ""


def root_domain(host):
    return re.sub(r"^www\.", "", host).lower()


def is_valid_email(email, site_host):
    if any(b in email for b in BLOCKLIST):
        return False
    if re.search(r"\.(png|jpe?g|gif|svg|webp|woff2?|ttf|ico|css|js)(\?|$)", email):
        return False
    if re.match(r"^[a-f0-9]{16,}@", email, re.IGNORECASE):
        return False
    if len(email) > 80:
        return False

    tld = tld_of(email)
    if tld in PLAUSIBLE_TLDS:
        return True

    if site_host:
        site_tld = root_domain(site_host).split(".")[-1]
        if tld == site_tld:
            return True
    return False


def extract(html, site_host):
    processed = preprocess(html)
    mailtos = set()
    plain = set()

    for m in re.finditer(r"mailto:([^\"'?\s>&]+)", processed, re.IGNORECASE):
        mailtos.add(m.group(1))
    for m in EMAIL_RE.finditer(processed):
        plain.add(m.group(0))

    def clean(e):
        return re.sub(r"[.,;:]+$", "", e.lower())

    mailto_list = [e for e in (clean(x) for x in mailtos) if is_valid_email(e, site_host)]
    plain_list = [e for e in (clean(x) for x in plain) if is_valid_email(e, site_host)]

    root = root_domain(site_host) if site_host else ""

    def same_domain(e):
        return bool(root) and (e.endswith("@" + root) or e.endswith("." + root))

    return (
        next((e for e in mailto_list if same_domain(e)), None)
        or next((e for e in plain_list if same_domain(e)), None)
        or (mailto_list[0] if mailto_list else None)
        or (plain_list[0] if plain_list else None)
        or ""
    )


def find_email(website):
    if not website:
        return ""
    base = urlparse(website)
    if not base.scheme or not base.hostname:
        return ""

    for path in PATHS:
        url = f"{base.scheme}://{base.hostname}{path}"
        html = fetch_text(url)
        if not html:
            continue
        email = extract(html, base.hostname)
        if email:
            return email
    return ""


def enrich_emails(leads, on_progress=console_progress):
    out = []
    for i, lead in enumerate(leads):
        email = find_email(lead.get("website"))
        on_progress(
            {
                "type": "item",
                "index": i + 1,
                "total": len(leads),
                "name": lead.get("name"),
                "status": email or "(sem email)",
            }
        )
        out.append({**lead, "email": email})
    return out
